src/diff.rs

```diff
@@ -182,91 +201,683 @@ fn diff_attributes(change_list: &mut ChangeListBuilder, old: &[Attribute], new:
+
+    // If that was all of the old children, then create and append the remaining
+    // new children and we're finished.
+    if shared_prefix_count == old.len() {
+        debug_assert!(
+            pushed,
+            "we handle the case of empty children before calling this method, so
+             `shared_prefix_count` must be greater than zero, and therefore we
+               must have pushed in the above loop"
+        );
+        change_list.pop();
+        create_and_append_children(
+            cached_set,
+            change_list,
+            registry,
+            &new[shared_prefix_count..],
+            cached_roots,
+        );
+        return KeyedPrefixResult::Finished;
+    }
+
+    // And if that was all of the new children, then remove all of the remaining
+    // old children and we're finished.
+    if shared_prefix_count == new.len() {
+        // Same as above.
+        debug_assert!(pushed);
+        change_list.pop_push_next_sibling();
+        change_list.remove_self_and_next_siblings();
+        return KeyedPrefixResult::Finished;
+    }
+
+    if pushed {
+        debug_assert!(shared_prefix_count > 0);
+        change_list.pop();
+    }
+    KeyedPrefixResult::MoreWorkToDo(shared_prefix_count)
+}
+
+// The most-general, expensive code path for keyed children diffing.
+//
+// We find the longest subsequence within `old` of children that are relatively
+// ordered the same way in `new` (via finding a longest-increasing-subsequence
+// of the old child's index within `new`). The children that are elements of
+// this subsequence will remain in place, minimizing the number of DOM moves we
+// will have to do.
+//
+// Upon entry to this function, the stack must be:
+//
+//     [... parent]
+//
+// Upon exit from this function, it will be restored to that same state.
+fn diff_keyed_middle(
+    cached_set: &CachedSet,
+    change_list: &mut ChangeListBuilder,
+    registry: &mut EventsRegistry,
+    old: &[Node],
+    new: &[Node],
+    cached_roots: &mut FxHashSet<CacheId>,
+    shared_prefix_count: usize,
+    shared_suffix_count: usize,
+    old_shared_suffix_start: usize,
+) {
+    // Should have already diffed the shared-key prefixes and suffixes.
+    debug_assert_ne!(new.first().map(|n| n.key()), old.first().map(|o| o.key()));
+    debug_assert_ne!(new.last().map(|n| n.key()), old.last().map(|o| o.key()));
+
+    // The algorithm below relies upon using `u32::MAX` as a sentinel
+    // value, so if we have that many new nodes, it won't work. This
+    // check is a bit academic (hence only enabled in debug), since
+    // wasm32 doesn't have enough address space to hold that many nodes
+    // in memory.
+    debug_assert!(new.len() < u32::MAX as usize);
+
+    // Map from each `old` node's key to its index within `old`.
+    let mut old_key_to_old_index = FxHashMap::default();
+    old_key_to_old_index.reserve(old.len());
+    old_key_to_old_index.extend(old.iter().enumerate().map(|(i, o)| (o.key(), i)));
```

Comment:
This brings up an interesting idea in the sense that it's the onus of the user to not reuse keys, right?

(should that be added to the `key` docs?

Reply:
Actually now I'm slightly confused, it looks like `u32::MAX` is used for non-present keys, so that just means that `u32::MAX` as a key in this map is overwritten multiple times? I guess though non-max values are supposed to be unique?

Reply:
Yeah I should add an assertion for this.